### PR TITLE
wxGenericListCtrl: Improve sorting

### PR DESCRIPTION
**Pull request by vvs31415**
Static data removed, eliminating contention.
Class wxListLineDataArray refactored not using dynarray

### CODE REVIEW

File: include/wx/generic/private/listctrl.h [hunk 1]
```diff
@@ -313,7 +313,18 @@ class wxListLineData
...
 };
 
-WX_DECLARE_OBJARRAY(wxListLineData, wxListLineDataArray);
+class wxListLineDataArray : public wxVector<wxListLineData*>
+{
+public:
+    void Clear()
+    {
+        for ( size_t n = 0; n < size(); ++n )
+            delete(operator[](n));
```
Comment (vadz) [inline include/wx/generic/private/listctrl.h]: This looks a bit weird, I'd write it either as `at(n)` (which is a bit less efficient due to the index check) or `(*this)[n]`.

File: src/generic/listctrl.cpp [hunk 2]
```diff
@@ -4480,9 +4478,11 @@ void wxListMainWindow::SortItems( wxListCtrlCompare fn, wxIntPtr data )
     HighlightAll(false);
     ResetCurrent();
 
-    list_ctrl_compare_func_2 = fn;
-    list_ctrl_compare_data = data;
-    m_lines.Sort( list_ctrl_compare_func_1 );
+    wxListLineComparator comparator;
```
Comment (vadz) [inline src/generic/listctrl.cpp]: IMHO it would make sense to give this struct a ctor taking `fn` and `data`.

**Review by vadz — COMMENTED** [review]
Thanks!

Globally looks good to me, but I would rephrase this commit message as "Switch to using wxVector instead of macro-based arrays" as most of the changes deal with this. Of course, the commit message could explain that this also allows to get rid of global variables.

I can merge this as is or I could wait a bit if you'd like to improve it further, please let me know.

File: include/wx/generic/private/listctrl.h (hunk 1 above)
Comment (vvs31415) [reply include/wx/generic/private/listctrl.h]: Fixed. This came from wxBaseObjectArray::DoEmpty()

**Comment by vvs31415** [comment]
This is like `struct wxInternalDataSort` in msw/listctrl.cpp.
I'd like to keep it as PlainOldData type to avoid any issues with templates.

File: src/generic/listctrl.cpp (hunk 2 above)
Comment (vvs31415) [reply src/generic/listctrl.cpp]: This is like struct wxInternalDataSort in msw/listctrl.cpp.
 I'd like to keep it as PlainOldData type to avoid any issues with templates.

**Comment by vvs31415** [comment]
Thanks! Please feel free to merge with any changes you deem appropriate.

**Comment by vadz** [comment]
I'm afraid this introduces memory leaks in `DeleteItem()`. Did you run the listctrl sample under a memory debugger to check if there are no new leaks now?

**Comment by vadz** [comment]
OK, there is definitely a leak, but fixing it is trivial, so I'll do it. Please keep in mind that `OBJARRAY`s own the items they contain for any future changes. Thanks!
